### ISSUE TITLE
DOSBox-X user guide

### ISSUE THREAD

**Issue by Wengier** [joncampbell123/dosbox-x#1412]
Thanks for your efforts on the project! Since I want to get more familiar with using the program, I was trying to find a DOSBox-X manual for end-users with complete usage information (e.g. supported commands, frequently asked questions). However, on the DOSBox-X site I could only find files such as README.md and README.Windows which provide information for developers (building and debugging etc), but I could not find the usage information from an end-user's perspective. There is also no such information on the dosbox-x.com website and the DOSBox-X Wiki. Then I confirmed with Edward Mendelson that no user guide exists yet indeed. However, I think in order to help end users to get familiar with DOSBox-X more easily, a user guide containing such information, e.g. internal & external commands, options, and sections such as FAQs will be very useful. I am also willing to contribute to such user guide when I have time.

**Comment by joncampbell123** [comment]
I appreciate it. Development will be slow for a few months due to getting tied up in my professional work, so this is the perfect time to do it. However this issue is a good place to discuss what DOSBox-X can do and I can make some time for that.

There's a lot to cover, go ahead and ask.

**Comment by Allofich** [comment]
I agree it would be probably be good to have this, even though most of the basic stuff is the same as regular DOSBox as far as I know.

One thing I would like to use this chance to mention is a minor feature I added in https://github.com/joncampbell123/dosbox-x/commit/13325c36fc7661f5ec06073a4f5ec3496f792170.

It was in response to a complaint that you couldn't use the `sm` command to set a value in memory to the hex values `AC` `AF` `CF` or `DF`. The problem was that these hex values are also the names of processor flags emulated by DOSBox-X, so if you tried to use these values, instead of setting the value of `AC` or whatever, instead the true/false value of the flag with that name would be used. For example, `sm 0000:0300 AC` would set memory location 0000:0300 to the 0 or 1 value of flag AC.

As a workaround I made it so you can put double quotes around the value to specify that you want to use that specific value, not the value of the flag with that name, so `sm 0000:0300 "AC"` can now be used to set memory location 0000:0300 to the hexidecimal value `AC`.

I noted this in the source code and added an entry to the change log, but with no user guide it could go unnoticed to anyone who isn't closely following development.

**Comment by joncampbell123** [comment]
It's funny how assemblers and C/C++ already solved the problem of hex vs decimal vs named by just requiring a 0x prefix, or for assemblers the 'h' suffix :)

**Comment by Wengier** [comment]
Thanks for the replies from both of you. In my opinion, even though DOSBox-X was originally forked from DOSBox, it has gradually evolved into its own project with its own focus and website thank to the efforts from you guys. Compared with original DOSBox, it has so many added or updated features, commands and options, along with its GUI menu that simply does not exist in original DOSBox. In such case, any existing documention for original DOSBox does not necessarily apply to DOSBox-X, and all the updates and new features in DOSBox-X are not covered in existing documents, not to mention that the differences between DOSBox and DOSBox-X will become even bigger as the project goes on. Therefore, I guess it will be necessary for DOSBox-X to have a user guide of its own.

As for the user guide itself, I think it can cover DOSBox-X's main focus and features, what DOSBox-X can do, how it differs from DOSBox (and possibly other DOS emulators), how to run and configure the program, the functions of its GUI menu items, the supported commands, command line options and dosbox.conf options, usage tips and remarks, and also sections like frequently asked questions. Of course both of you will decide how it will eventually look like. In any case, I think this will be very helpful to get users started.

**Comment by Allofich** [comment]
> It's funny how assemblers and C/C++ already solved the problem of hex vs decimal vs named by just requiring a 0x prefix, or for assemblers the 'h' suffix :)

Btw @joncampbell123 if you prefer a different solution than what I did, please feel free to change it.

**Comment by Wengier** [comment]
@Allofich Yes, I think more can be said about the dosbox.conf options. Some comments in that file are relatively brief to keep the file size small I guess. But longer (and clear) explanations will probably be desired for the purpose of a user guide.

@joncampbell123 I agree there's a lot to cover. And I was about to start writing in an offline text-file named README.TXT on my computer as I did with other projects before. But now I think about it, perhaps we can make use of the existing DOSBox-X Wiki system (https://github.com/joncampbell123/dosbox-x/wiki) to cooperate about this? There is not really much information on the DOSBox-X Wiki right now. But I think it is completely possible to turn it into an online DOSBox-X user guide if we can cooperate this way (I contributed to other Wiki before, so Wiki-style writing is ok for me). How do you think about this? Thanks!

**Comment by Wengier** [comment]
@joncampbell123 You can let me know when the Wiki content is ready to such an expansion. No need to hurry, but I am happy to contribute then.

**Comment by joncampbell123** [comment]
You can git clone the wiki and work locally if you like:

git clone https://github.com/joncampbell123/dosbox-x.wiki.git dosbox-x-wiki

**Comment by Wengier** [comment]
@joncampbell123 I had cloned the wiki and then made the DOSBox-X user guide page under my GitHub account:

https://github.com/Wengier/dosbox-x-wiki/wiki

It currently has the following sections:

1. Introduction

2. DOSBox-X's Main Focus

3. What DOSBox-X Can Do

4. DOSBox-X's Feature Highlights

5. Running and Configuring DOSBox-X

6. DOSBox-X's GUI Menus

7. DOSBox-X's Special Keys

8. DOSBox-X's Supported Commands

9. DOSBox-X's Command-line Options

10. Compatibility

11. Frequently Asked Questions (FAQ)

It took me some time to add contents to this page as it appears now, but of course there is still more information that can be added to the guide. For example, more questions and answers can be added to the FAQ section, but of course as the project maintainer you will decide which questions and answers will appear in it. Hope this will be helpful for DOSBox-X users.

**Comment by Wengier** [comment]
@joncampbell123 and @Allofich: The user guide is currently still a draft, so if you find any issues in it please let me know, or you can just copy and modify it directly as you wish. Hope other users will benefit from such a guide.

**Comment by joncampbell123** [comment]
Running a ROM BIOS image is still experimental as well. At this time it's only good for custom code and asm assembly experiments. :)

Also, F11 is the host key only on Windows. F12 is the host key on any other platform. The reason is that F12 is used internally by Windows for debugging functions.

I added Pause with Interrupts Enabled because most Demoscene stuff and games use interrupts to play music and it's a great way to pause the game/demo without pausing the music. It's also a good way to hear the entire MOD in a Demoscene production when the demo exits long before the MOD has time to loop.

**Comment by joncampbell123** [comment]
The normal/speed up/speed down commands allow you to control emulation speed relative to realtime. For example if you want to play a game at 50% normal speed, you can slow it down. Normal speed restores normal emulation.

**Comment by joncampbell123** [comment]
Direct3D shader support is (obviously) specific to Windows. There is no visual indication if the shader fails to load, it's written to the log file. If you want more immediate feedback on success or failure, use the menu to show the console which will also show the reason for the shader failure.

**Comment by joncampbell123** [comment]
The Drive menu is an incomplete recreation of the Drive menu that used to exist when the menu was a Windows-specific *.rc menu resource.

**Comment by joncampbell123** [comment]
F11/F12 is the "Host key" which is implemented as a modifier alongside what is normally ctrl, shift, alt. The host key can be redefined in the mapper as needed, if you don't like using F11/F12.

**Comment by joncampbell123** [comment]
DX-CAPTURE is a built-in shell command that runs a command with intent to capture.

It starts AVI/WAV/etc. capture (according to what you specify), then runs the program, then automatically stops capture when the program exits (with a pause after exit). I use it to make Demoscene captures and to make sure the capture stops when it exits.

The command name is deliberately longer than 8 chars so that there is no conflict with external COM/EXE executables that are limited to 8.3 filenames.

**Comment by joncampbell123** [comment]
So far at quick glance the rest looks good. 👍

**Comment by Wengier** [comment]
Thanks! I have updated the user guide wiki content using these information, i.e. explanations for the host key, normal/speed up/speed down, Direct3D shader support, the "Pause with Interrupts Enabled" feature, etc. I have added a "Note" with the detailed information you provided to some of DOSBox-X's commands (e.g. BOOT and DX-CAPTURE). Thanks again for this!

**Comment by joncampbell123** [comment]
Other descriptions:

VFRCRATE is a tool that can be used to force video emulation to a specific refresh rate (or turn off the forced rate). I originally wrote it to run demoscene stuff at 59.94Hz (NTSC) so that I didn't have to do any frame blending to author to DVD (something I was doing at the time I originally wrote the feature in 2008 as a patch against SVN). It can also be used for development and testing to simulate a PC who's refresh rate is locked in hardware, such as what happens when running a DOS program on a laptop when video is directed to the LCD display. Even though standard VGA is 60Hz or 70Hz, laptops will lock the refresh rate to 60Hz when sending video to the internal display.

**Comment by joncampbell123** [comment]
Some of the builtin tools like HEXMEM and DSXMENU are open source. Source code is in the related DOSLIB project. HEXMEM was specifically written as a way to poke around the addressable memory available to the CPU and to show how a 16-bit DOS program can access extended memory, including flat real mode, and the 286 reset vector trick for 286 systems. There is also code to access memory above 4GB if the CPU supports 64-bit long mode or the PAE page table extensions.

**Comment by joncampbell123** [comment]
RE-DOS sends a signal to re-boot the DOS kernel, without rebooting the guest system.

**Comment by joncampbell123** [comment]
In addition to RESCAN, the MOUNT command accepts a -nocachedir option to flush the cache often so that RESCAN is not needed, for users who want DIR to always show the host folder's contents.

**Comment by joncampbell123** [comment]
VESAMOED is a utility to allow scripts or the user to add/modify/delete VESA BIOS modes. It was originally written because some old DOS games or demoscene productions, especially those shipped with a UNIVBE binary, assumed video mode numbers instead of enumerating like they should. It can also be used to rearrange VESA BIOS modes for retro developers who want to make sure their code works properly no matter what strange VESA BIOS their code runs into on real hardware.

Because of limitations in DOSBox-X SVGA emulation and the render scaler architecture, the maximum resolution possible resolution is 1920x1440.

**Comment by joncampbell123** [comment]
NMITEST generates a non-maskable interrupt. This is a debugging tool to test that it and the interrupt handler work properly.

Currently the only use of the NMI is PCjr emulation which receives an NMI every time you press a key on the keyboard.

**Comment by joncampbell123** [comment]
Please indicate when you are ready for your changes to be merged back into the Dosbox-X wiki repo.

**Comment by joncampbell123** [comment]
The idea of the multi track AVI capture by the way is to allow you to capture each mixer channel separately in one file, instead of mixed together. So the AVI file would have one audio track for PC speaker, another for the Sound Blaster DSP, another for the OPL2/OPL3, and another for the Gravis Ultrasound. It makes it possible in video production to focus on the digitized speech and fade out the music as needed, or remove the sound effects and PC speaker sounds and leave only the music.

Unfortunately Adobe Premiere doesn't recognize multi-track AVI files like that, which is why there is a script in DOSBox-X's source tree to use FFMPEG to dump each one out into a separate WAV file.

**Comment by joncampbell123** [comment]
Of course that flexibility is lost in games that do their own audio mixing in software or through the Gravis Ultrasound, but it is useful for games that use the OPL2/3 for music and Sound Blaster DSP for sound.

**Comment by joncampbell123** [comment]
The capture format selection allows you to choose between the traditional AVI (ZMBV) format already used in DOSBox SVN and it's forks, or an MPEG-2 transport stream that uses H.264 video and AAC audio. The MPEG-2 TS option requires DOSBox-X to be compiled with FFMPEG support, otherwise the option will not be available.

Contrary to DOS enthusiast fears when I added MPEG-2 TS as a format, I will never remove the ZMBV AVI capture mode.

The idea behind the MPEG-2 TS format is that it is possible to support video refresh rate changes mid-stream (which the AVI format cannot do). If you are capturing a game or demoscene production that changes video modes often, the MPEG-2 TS format is preferable because it means you can make a capture of the demo without having to stitch together multiple AVI files that would normally happen, a new AVI file every video mode change.

I'm told that's not explicitly supported, but so far VLC player and FFMPEG have absolutely no problem with it.

**Comment by Wengier** [comment]
> Please indicate when you are ready for your changes to be merged back into the Dosbox-X wiki repo.

@joncampbell123 I have so far incorporated all the information you provided up to this post (including the details for VFRCRATE, HEXMEM, DSXMENU, RE-DOS, MOUNT, VESAMOED, and NMITEST commands). I think I am ready for the changes to be merged back into the DOSBox-X wiki now.

**Comment by joncampbell123** [comment]
Done. Check the DOSBox-X wiki.

**Comment by Wengier** [comment]
The user guide at the DOSBox-X wiki (https://github.com/joncampbell123/dosbox-x/wiki) looks good. Thanks for all of your work on the project! 👍
